# Cours 4


## Session 3

Reprise de la session 3 : 
- Reprise de l'API
- Regex est plus fragile que BeautifulSoup, mais si le site change, on peut jeter la version BeautifulSoup. 

## Session 4

Préparation des données

- trouver/supprimer les données dupliquées : df.duplicated() / df.drop_duplicates()
- renommer des colonnes : df.rename(columns={...))
- trouver les NaN : df.isnull() / df.notnull() / df.dropna()
- travail sur les chaines : series.str.extract(), series.str.contains(), series.get_dummies()
- mapping : series.map()
- changer le type d'une série (cast) : df.astype(type) / pd.to_numeric() / pd.to_datetime()
- remplacer n'importe quelle valeur : df.replace({...})
- remplacer les NaN : df.fillna(), series.combine_first()

In [78]:
# imports
import numpy as np
import pandas as pd


#### Chargement des données 

In [79]:
# load data
df = pd.read_csv('people.csv')
df

id  first_name  last_name                 email address  gender age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1    27570       Ruthi       Ross              rross3w@sohu.com  Female  57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female  54   
..     ...         ...        ...                           ...     ...  ..   
204  27465     Adelind   Christin          achristinz@blogs.com  Female  30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female  28   
206  27521      Moritz    Issacof          missacof2j@wired.com    Male  21   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com       F  32   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl    Male  34   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1    $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2    €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3    €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4    $13.30   14.772557  45.160472         NaN   2018/08/13    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019/06/04    False   
205  €37.15  132.702111  33.762296         NaN   2017/11/22    False   
206  $91.97   72.311099  31.263396  0298949280   2018/08/20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018/10/25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017/06/05    False   

        last_seen                     address           preference  
0    1.559566e+09   Palkovice, Czech Republic  entrée/plat/dessert  
1    1.567165e+09            Siewierz, Poland  entrée/plat/dessert  
2    1.553692e+09               Longbo, China  entrée/plat/dessert  
3    1.552349e+09        Potulando, Indonesia          entrée/plat  
4    1.543383e+09             Bribir, Croatia  entrée/plat/dessert  
..            ...                         ...                  ...  
204  1.559606e+09            Znamenka, Russia  entrée/plat/dessert  
205  1.532870e+09                  Iyo, Japan  entrée/plat/dessert  
206  1.549817e+09        Jhang Sadr, Pakistan  entrée/plat/dessert  
207  1.540426e+09  Mantingantengah, Indonesia  entrée/plat/dessert  
208  1.540441e+09          Indaiatuba, Brazil  entrée/plat/dessert  

[209 rows x 15 columns]

In [80]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209 entries, 0 to 208
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             209 non-null    int64  
 1   first_name     207 non-null    object 
 2   last_name      207 non-null    object 
 3   email address  203 non-null    object 
 4   gender         207 non-null    object 
 5   age            207 non-null    object 
 6   money          190 non-null    object 
 7   lon            207 non-null    float64
 8   lat            207 non-null    float64
 9   phone          83 non-null     object 
 10  registration   207 non-null    object 
 11  inactive       207 non-null    object 
 12  last_seen      190 non-null    float64
 13  address        207 non-null    object 
 14  preference     207 non-null    object 
dtypes: float64(3), int64(1), object(11)
memory usage: 24.6+ KB


Le fichier est un petit fichier avec uniquement 109 lignes. 

On va travailler colonne sur colonne. On va faire du data quality assessment => NETTOYAGE des données à la main 

#### Duplications
- duplicated() : True ou False selon si une ligne est dupliquée
- drop_duplicates() : suppression des lignes dupliquées

In [81]:
# lignes dupliquées
df.duplicated().value_counts()

False    205
True       4
dtype: int64

Ici il y a 4 lignes qui sont dupliquées. 

Les fonctions duplicates renvoient des copies du dataframe. 


In [82]:
#Suppresion des lignes dupliquées
df = df.drop_duplicates() 
df

id  first_name  last_name                 email address  gender age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1    27570       Ruthi       Ross              rross3w@sohu.com  Female  57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female  54   
..     ...         ...        ...                           ...     ...  ..   
204  27465     Adelind   Christin          achristinz@blogs.com  Female  30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female  28   
206  27521      Moritz    Issacof          missacof2j@wired.com    Male  21   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com       F  32   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl    Male  34   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1    $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2    €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3    €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4    $13.30   14.772557  45.160472         NaN   2018/08/13    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019/06/04    False   
205  €37.15  132.702111  33.762296         NaN   2017/11/22    False   
206  $91.97   72.311099  31.263396  0298949280   2018/08/20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018/10/25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017/06/05    False   

        last_seen                     address           preference  
0    1.559566e+09   Palkovice, Czech Republic  entrée/plat/dessert  
1    1.567165e+09            Siewierz, Poland  entrée/plat/dessert  
2    1.553692e+09               Longbo, China  entrée/plat/dessert  
3    1.552349e+09        Potulando, Indonesia          entrée/plat  
4    1.543383e+09             Bribir, Croatia  entrée/plat/dessert  
..            ...                         ...                  ...  
204  1.559606e+09            Znamenka, Russia  entrée/plat/dessert  
205  1.532870e+09                  Iyo, Japan  entrée/plat/dessert  
206  1.549817e+09        Jhang Sadr, Pakistan  entrée/plat/dessert  
207  1.540426e+09  Mantingantengah, Indonesia  entrée/plat/dessert  
208  1.540441e+09          Indaiatuba, Brazil  entrée/plat/dessert  

[205 rows x 15 columns]

In [83]:
df.duplicated().value_counts()

False    205
dtype: int64

In [84]:
df = df.rename(columns={'email address': 'email'})
df

id  first_name  last_name                         email  gender age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1    27570       Ruthi       Ross              rross3w@sohu.com  Female  57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female  54   
..     ...         ...        ...                           ...     ...  ..   
204  27465     Adelind   Christin          achristinz@blogs.com  Female  30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female  28   
206  27521      Moritz    Issacof          missacof2j@wired.com    Male  21   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com       F  32   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl    Male  34   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1    $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2    €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3    €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4    $13.30   14.772557  45.160472         NaN   2018/08/13    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019/06/04    False   
205  €37.15  132.702111  33.762296         NaN   2017/11/22    False   
206  $91.97   72.311099  31.263396  0298949280   2018/08/20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018/10/25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017/06/05    False   

        last_seen                     address           preference  
0    1.559566e+09   Palkovice, Czech Republic  entrée/plat/dessert  
1    1.567165e+09            Siewierz, Poland  entrée/plat/dessert  
2    1.553692e+09               Longbo, China  entrée/plat/dessert  
3    1.552349e+09        Potulando, Indonesia          entrée/plat  
4    1.543383e+09             Bribir, Croatia  entrée/plat/dessert  
..            ...                         ...                  ...  
204  1.559606e+09            Znamenka, Russia  entrée/plat/dessert  
205  1.532870e+09                  Iyo, Japan  entrée/plat/dessert  
206  1.549817e+09        Jhang Sadr, Pakistan  entrée/plat/dessert  
207  1.540426e+09  Mantingantengah, Indonesia  entrée/plat/dessert  
208  1.540441e+09          Indaiatuba, Brazil  entrée/plat/dessert  

[205 rows x 15 columns]

#### Analyse des données manquantes 

In [85]:
np.nan

nan

In [86]:
print(type(np.nan))

<class 'float'>


In [87]:
np.sqrt(np.nan)

nan

In [88]:
np.nan == np.nan

False

In [89]:
np.NINF, np.inf

(-inf, inf)

In [90]:
print (np.nan>1)
print(np.nan<1)
print(np.inf>1)
print(np.inf + np.inf)
print(np.inf +2)

False
False
True
inf
inf


In [91]:
print(np.inf + np.NINF)

nan


In [92]:
np.inf >np.inf

False

In [93]:
np.inf + np.inf > np.inf

False

#### Tests sur les données manquantes 

In [94]:
df.loc[df['first_name']!= np.nan]

id  first_name  last_name                         email  gender age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1    27570       Ruthi       Ross              rross3w@sohu.com  Female  57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female  54   
..     ...         ...        ...                           ...     ...  ..   
204  27465     Adelind   Christin          achristinz@blogs.com  Female  30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female  28   
206  27521      Moritz    Issacof          missacof2j@wired.com    Male  21   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com       F  32   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl    Male  34   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1    $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2    €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3    €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4    $13.30   14.772557  45.160472         NaN   2018/08/13    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019/06/04    False   
205  €37.15  132.702111  33.762296         NaN   2017/11/22    False   
206  $91.97   72.311099  31.263396  0298949280   2018/08/20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018/10/25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017/06/05    False   

        last_seen                     address           preference  
0    1.559566e+09   Palkovice, Czech Republic  entrée/plat/dessert  
1    1.567165e+09            Siewierz, Poland  entrée/plat/dessert  
2    1.553692e+09               Longbo, China  entrée/plat/dessert  
3    1.552349e+09        Potulando, Indonesia          entrée/plat  
4    1.543383e+09             Bribir, Croatia  entrée/plat/dessert  
..            ...                         ...                  ...  
204  1.559606e+09            Znamenka, Russia  entrée/plat/dessert  
205  1.532870e+09                  Iyo, Japan  entrée/plat/dessert  
206  1.549817e+09        Jhang Sadr, Pakistan  entrée/plat/dessert  
207  1.540426e+09  Mantingantengah, Indonesia  entrée/plat/dessert  
208  1.540441e+09          Indaiatuba, Brazil  entrée/plat/dessert  

[205 rows x 15 columns]

In [95]:
df.loc[df['first_name'].isnull()]

id first_name last_name email gender  age money  lon  lat phone  \
96   27565        NaN       NaN   NaN    NaN  NaN   NaN  NaN  NaN   NaN   
200  27630        NaN       NaN   NaN    NaN  NaN   NaN  NaN  NaN   NaN   

    registration inactive  last_seen address preference  
96           NaN      NaN        NaN     NaN        NaN  
200          NaN      NaN        NaN     NaN        NaN

In [96]:
df.loc[df['first_name'].isna()]

id first_name last_name email gender  age money  lon  lat phone  \
96   27565        NaN       NaN   NaN    NaN  NaN   NaN  NaN  NaN   NaN   
200  27630        NaN       NaN   NaN    NaN  NaN   NaN  NaN  NaN   NaN   

    registration inactive  last_seen address preference  
96           NaN      NaN        NaN     NaN        NaN  
200          NaN      NaN        NaN     NaN        NaN

isna et isnull donne la même chose. 

In [97]:
df.isnull()

id  first_name  last_name  email  gender    age  money    lon    lat  \
0    False       False      False  False   False  False  False  False  False   
1    False       False      False  False   False  False  False  False  False   
2    False       False      False  False   False  False  False  False  False   
3    False       False      False  False   False  False  False  False  False   
4    False       False      False  False   False  False  False  False  False   
..     ...         ...        ...    ...     ...    ...    ...    ...    ...   
204  False       False      False  False   False  False  False  False  False   
205  False       False      False  False   False  False  False  False  False   
206  False       False      False  False   False  False  False  False  False   
207  False       False      False  False   False  False  False  False  False   
208  False       False      False  False   False  False  False  False  False   

     phone  registration  inactive  last_seen  address  preference  
0    False         False     False      False    False       False  
1     True         False     False      False    False       False  
2     True         False     False      False    False       False  
3     True         False     False      False    False       False  
4     True         False     False      False    False       False  
..     ...           ...       ...        ...      ...         ...  
204   True         False     False      False    False       False  
205   True         False     False      False    False       False  
206  False         False     False      False    False       False  
207  False         False     False      False    False       False  
208  False         False     False      False    False       False  

[205 rows x 15 columns]

In [98]:
df.loc[df.isnull().any(axis =1)] # on veut connaitre toutes les lignes qui ont au moins un NaN

id  first_name  last_name                         email  gender  age  \
1    27570       Ruthi       Ross              rross3w@sohu.com  Female   57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male   22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F   64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female   54   
6    27539     Dulcine  Stopforth       dstopforth31@forbes.com  Female   22   
..     ...         ...        ...                           ...     ...  ...   
200  27630         NaN        NaN                           NaN     NaN  NaN   
201  27519     Leonora      Guild           lguild2h@census.gov  Female   54   
203  27612    Torrance    Sivills       tsivills52@mapquest.com    Male   24   
204  27465     Adelind   Christin          achristinz@blogs.com  Female   30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female   28   

      money         lon        lat phone registration inactive     last_seen  \
1    $20.37   19.230220  50.466575   NaN   2018/10/23    False  1.567165e+09   
2    €32,99  118.831081  24.984813   NaN   2018/12/30    False  1.553692e+09   
3    €98,93  121.648987  -8.844744   NaN   2019/03/12    False  1.552349e+09   
4    $13.30   14.772557  45.160472   NaN   2018/08/13    False  1.543383e+09   
6    €12.60  -35.587670  -6.255393   NaN   2017/12/13    False  1.547567e+09   
..      ...         ...        ...   ...          ...      ...           ...   
200     NaN         NaN        NaN   NaN          NaN      NaN           NaN   
201  €54.57  122.195048   7.552335   NaN   2018/12/17    False  1.545005e+09   
203  $36.43  -91.148259  15.027043   NaN   2018/04/03    False  1.549641e+09   
204  €81,84   41.427853  52.415968   NaN   2019/06/04    False  1.559606e+09   
205  €37.15  132.702111  33.762296   NaN   2017/11/22    False  1.532870e+09   

                              address           preference  
1                    Siewierz, Poland  entrée/plat/dessert  
2                       Longbo, China  entrée/plat/dessert  
3                Potulando, Indonesia          entrée/plat  
4                     Bribir, Croatia  entrée/plat/dessert  
6                    Serrinha, Brazil  entrée/plat/dessert  
..                                ...                  ...  
200                               NaN                  NaN  
201              Siraway, Philippines  entrée/plat/dessert  
203  Santa Cruz del Quiché, Guatemala  entrée/plat/dessert  
204                  Znamenka, Russia  entrée/plat/dessert  
205                        Iyo, Japan  entrée/plat/dessert  

[134 rows x 15 columns]

Dans le dataframe de 209 lignes, il y a 134 lignes qui sont composées d'au moins 1 NaN. 
Ce n'est donc pas une bonne méthode.

In [99]:
df1 = df.dropna()

In [100]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 205 entries, 0 to 208
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            205 non-null    int64  
 1   first_name    203 non-null    object 
 2   last_name     203 non-null    object 
 3   email         199 non-null    object 
 4   gender        203 non-null    object 
 5   age           203 non-null    object 
 6   money         186 non-null    object 
 7   lon           203 non-null    float64
 8   lat           203 non-null    float64
 9   phone         80 non-null     object 
 10  registration  203 non-null    object 
 11  inactive      203 non-null    object 
 12  last_seen     186 non-null    float64
 13  address       203 non-null    object 
 14  preference    203 non-null    object 
dtypes: float64(3), int64(1), object(11)
memory usage: 25.6+ KB


In [101]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 71 entries, 0 to 208
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            71 non-null     int64  
 1   first_name    71 non-null     object 
 2   last_name     71 non-null     object 
 3   email         71 non-null     object 
 4   gender        71 non-null     object 
 5   age           71 non-null     object 
 6   money         71 non-null     object 
 7   lon           71 non-null     float64
 8   lat           71 non-null     float64
 9   phone         71 non-null     object 
 10  registration  71 non-null     object 
 11  inactive      71 non-null     object 
 12  last_seen     71 non-null     float64
 13  address       71 non-null     object 
 14  preference    71 non-null     object 
dtypes: float64(3), int64(1), object(11)
memory usage: 8.9+ KB


In [102]:
#Les lignes nulles qui nous intéressent sont celles dont le prénom est nul. 
#on veut les supprimer

df = df.dropna(subset=['first_name']) #on décide de supprimer les lignes dont le prénom est nul
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 203 entries, 0 to 208
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            203 non-null    int64  
 1   first_name    203 non-null    object 
 2   last_name     203 non-null    object 
 3   email         199 non-null    object 
 4   gender        203 non-null    object 
 5   age           203 non-null    object 
 6   money         186 non-null    object 
 7   lon           203 non-null    float64
 8   lat           203 non-null    float64
 9   phone         80 non-null     object 
 10  registration  203 non-null    object 
 11  inactive      203 non-null    object 
 12  last_seen     186 non-null    float64
 13  address       203 non-null    object 
 14  preference    203 non-null    object 
dtypes: float64(3), int64(1), object(11)
memory usage: 25.4+ KB


on a bien 203 entrées. 

In [103]:
df['full_name'] = df['first_name'] + ' ' + df['last_name']

<ipython-input-103-1296497f99fa>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['full_name'] = df['first_name'] + ' ' + df['last_name']


In [104]:
df

id  first_name  last_name                         email  gender age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1    27570       Ruthi       Ross              rross3w@sohu.com  Female  57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female  54   
..     ...         ...        ...                           ...     ...  ..   
204  27465     Adelind   Christin          achristinz@blogs.com  Female  30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female  28   
206  27521      Moritz    Issacof          missacof2j@wired.com    Male  21   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com       F  32   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl    Male  34   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1    $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2    €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3    €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4    $13.30   14.772557  45.160472         NaN   2018/08/13    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019/06/04    False   
205  €37.15  132.702111  33.762296         NaN   2017/11/22    False   
206  $91.97   72.311099  31.263396  0298949280   2018/08/20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018/10/25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017/06/05    False   

        last_seen                     address           preference  \
0    1.559566e+09   Palkovice, Czech Republic  entrée/plat/dessert   
1    1.567165e+09            Siewierz, Poland  entrée/plat/dessert   
2    1.553692e+09               Longbo, China  entrée/plat/dessert   
3    1.552349e+09        Potulando, Indonesia          entrée/plat   
4    1.543383e+09             Bribir, Croatia  entrée/plat/dessert   
..            ...                         ...                  ...   
204  1.559606e+09            Znamenka, Russia  entrée/plat/dessert   
205  1.532870e+09                  Iyo, Japan  entrée/plat/dessert   
206  1.549817e+09        Jhang Sadr, Pakistan  entrée/plat/dessert   
207  1.540426e+09  Mantingantengah, Indonesia  entrée/plat/dessert   
208  1.540441e+09          Indaiatuba, Brazil  entrée/plat/dessert   

               full_name  
0        Leandra Pabelik  
1             Ruthi Ross  
2         Silas Stourton  
3       Roxine Pettecrew  
4       Margaux Gowanson  
..                   ...  
204     Adelind Christin  
205  Petronella Pickance  
206       Moritz Issacof  
207   Quintana Foulstone  
208     Trueman De Haven  

[203 rows x 16 columns]

In [105]:
df[['city','country']] = df['address'].str.extract('(.*),(.*)')

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:2963: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [106]:
df

id  first_name  last_name                         email  gender age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1    27570       Ruthi       Ross              rross3w@sohu.com  Female  57   
2    27572       Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4    27558     Margaux   Gowanson                nope@thankyou.  Female  54   
..     ...         ...        ...                           ...     ...  ..   
204  27465     Adelind   Christin          achristinz@blogs.com  Female  30   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu  Female  28   
206  27521      Moritz    Issacof          missacof2j@wired.com    Male  21   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com       F  32   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl    Male  34   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1    $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2    €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3    €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4    $13.30   14.772557  45.160472         NaN   2018/08/13    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019/06/04    False   
205  €37.15  132.702111  33.762296         NaN   2017/11/22    False   
206  $91.97   72.311099  31.263396  0298949280   2018/08/20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018/10/25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017/06/05    False   

        last_seen                     address           preference  \
0    1.559566e+09   Palkovice, Czech Republic  entrée/plat/dessert   
1    1.567165e+09            Siewierz, Poland  entrée/plat/dessert   
2    1.553692e+09               Longbo, China  entrée/plat/dessert   
3    1.552349e+09        Potulando, Indonesia          entrée/plat   
4    1.543383e+09             Bribir, Croatia  entrée/plat/dessert   
..            ...                         ...                  ...   
204  1.559606e+09            Znamenka, Russia  entrée/plat/dessert   
205  1.532870e+09                  Iyo, Japan  entrée/plat/dessert   
206  1.549817e+09        Jhang Sadr, Pakistan  entrée/plat/dessert   
207  1.540426e+09  Mantingantengah, Indonesia  entrée/plat/dessert   
208  1.540441e+09          Indaiatuba, Brazil  entrée/plat/dessert   

               full_name             city          country  
0        Leandra Pabelik        Palkovice   Czech Republic  
1             Ruthi Ross         Siewierz           Poland  
2         Silas Stourton           Longbo            China  
3       Roxine Pettecrew        Potulando        Indonesia  
4       Margaux Gowanson           Bribir          Croatia  
..                   ...              ...              ...  
204     Adelind Christin         Znamenka           Russia  
205  Petronella Pickance              Iyo            Japan  
206       Moritz Issacof       Jhang Sadr         Pakistan  
207   Quintana Foulstone  Mantingantengah        Indonesia  
208     Trueman De Haven       Indaiatuba           Brazil  

[203 rows x 18 columns]

In [107]:
df.nunique() #fabrique un nunique pour chaque colonnes
#i.e. donne les nombres d'uniques. 

id              201
first_name      195
last_name       201
email           197
gender            4
age              54
money           184
lon             199
lat             199
phone            80
registration    183
inactive          1
last_seen       184
address         200
preference        5
full_name       201
city            200
country          60
dtype: int64

gender = 4 ??? 
On est dans un ancien fichier: on s'intéresse donc à la colonne gender... 

In [108]:
df['gender'].unique()

array(['Female', 'Male', 'F', 'M'], dtype=object)

Le fichier n'est donc pas normalisé. 

In [109]:
df['gender'].value_counts()

Male      89
Female    79
M         22
F         13
Name: gender, dtype: int64

In [110]:
df.head()

id first_name  last_name                         email  gender age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu  Female  63   
1  27570      Ruthi       Ross              rross3w@sohu.com  Female  57   
2  27572      Silas   Stourton  silas.stourton3y@answers.com    Male  22   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com       F  64   
4  27558    Margaux   Gowanson                nope@thankyou.  Female  54   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1  $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2  €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3  €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4  $13.30   14.772557  45.160472         NaN   2018/08/13    False   

      last_seen                    address           preference  \
0  1.559566e+09  Palkovice, Czech Republic  entrée/plat/dessert   
1  1.567165e+09           Siewierz, Poland  entrée/plat/dessert   
2  1.553692e+09              Longbo, China  entrée/plat/dessert   
3  1.552349e+09       Potulando, Indonesia          entrée/plat   
4  1.543383e+09            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country  
0   Leandra Pabelik  Palkovice   Czech Republic  
1        Ruthi Ross   Siewierz           Poland  
2    Silas Stourton     Longbo            China  
3  Roxine Pettecrew  Potulando        Indonesia  
4  Margaux Gowanson     Bribir          Croatia

In [112]:
#pour traiter, il faut mapper
#creation d'un dictionnaire avec le mapping que l'on souhaite

mapping = {'Female': 'F', 'F' : 'F', 'M' : 'M', 'Male': 'M'}
df['gender'] = df['gender'].map(mapping)
df.head()

<ipython-input-112-5c37dffff57d>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'] = df['gender'].map(mapping)


id first_name  last_name                         email gender age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu      F  63   
1  27570      Ruthi       Ross              rross3w@sohu.com      F  57   
2  27572      Silas   Stourton  silas.stourton3y@answers.com      M  22   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64   
4  27558    Margaux   Gowanson                nope@thankyou.      F  54   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019/04/16    False   
1  $20.37   19.230220  50.466575         NaN   2018/10/23    False   
2  €32,99  118.831081  24.984813         NaN   2018/12/30    False   
3  €98,93  121.648987  -8.844744         NaN   2019/03/12    False   
4  $13.30   14.772557  45.160472         NaN   2018/08/13    False   

      last_seen                    address           preference  \
0  1.559566e+09  Palkovice, Czech Republic  entrée/plat/dessert   
1  1.567165e+09           Siewierz, Poland  entrée/plat/dessert   
2  1.553692e+09              Longbo, China  entrée/plat/dessert   
3  1.552349e+09       Potulando, Indonesia          entrée/plat   
4  1.543383e+09            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country  
0   Leandra Pabelik  Palkovice   Czech Republic  
1        Ruthi Ross   Siewierz           Poland  
2    Silas Stourton     Longbo            China  
3  Roxine Pettecrew  Potulando        Indonesia  
4  Margaux Gowanson     Bribir          Croatia

Attention, ici si on ne met pas 'F' et 'M' dans le mapping, remplace les F et M initiales du dataframe par NaN

In [113]:
len(df.loc[df['gender'].isnull()])

0

In [114]:
df['gender'].value_counts()

M    111
F     92
Name: gender, dtype: int64

In [ ]:
#pour compléter le genre => utilisation d'une API gender. 

In [115]:
# load data
df1 = pd.read_csv('people.csv')
df1.loc[df1['gender'].isnull()]

id first_name last_name email address gender  age money  lon  lat  \
96   27565        NaN       NaN           NaN    NaN  NaN   NaN  NaN  NaN   
200  27630        NaN       NaN           NaN    NaN  NaN   NaN  NaN  NaN   

    phone registration inactive  last_seen address preference  
96    NaN          NaN      NaN        NaN     NaN        NaN  
200   NaN          NaN      NaN        NaN     NaN        NaN

Quand on travaille sur des dictionnaires incomplets, il faut travailler avec apply et pas map, car on peut écraser de la données. 

ATTENTION : ici reprendre ce qu'a fait le professeur 
    avec le comportement de map()
    et l'action avec un dictionnaire partiel

La fonction map nécessite un dictionnaire complet. 

#### Analyse de l'âge 

In [117]:
df['age'].value_counts()

54       11
70        7
24        7
31        6
42        6
47        6
32        6
63        6
44        6
33        6
55        5
65        5
20        5
38        5
28        5
57        5
27        4
35        4
66        4
41        4
49        4
45        4
no        4
48        4
43        4
60        4
68        4
61        4
62        4
22        4
52        4
25        3
56        3
26        3
59        3
21        3
30        3
64        3
23        2
51        2
nope      2
sorry     2
36        2
40        2
67        2
34        2
39        2
53        1
50        1
29        1
58        1
37        1
69        1
46        1
Name: age, dtype: int64

In [119]:
df.loc[df['age'].str.contains('[^0-9]'), 'age'].value_counts()
#le chapeau ^ signifie que l'on ne veut pas les éléments numériques de 0 à 9

no       4
sorry    2
nope     2
Name: age, dtype: int64

astype() : casting de type

si on veut gérer les erreurs on utilise le pandas.to_numeric() 

In [120]:
df.astype(int) #ne peut pas se faire

ValueError: Cannot convert non-finite values (NA or inf) to integer

In [121]:
# traitement de l'âge
df['age'] = pd.to_numeric(df['age'], errors='coerce')

#ici le coerce transforme les erreurs en NaN

<ipython-input-121-b62c39e6a17b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age'] = pd.to_numeric(df['age'], errors='coerce')


In [123]:
df.loc[df['age'].isnull()]

id first_name   last_name                             email gender  \
19   27507       Tate    Feldmann  tfeldmann25@networksolutions.com      M   
50   27483      Hanny      Ballin  hballin1h@networkadvertising.org      F   
59   27478      Kinna    Ffoulkes           kffoulkes1c@archive.org      F   
64   27506     Brooke    Edwicker       bedwicker24@hugedomains.com      M   
71   27430   Cordelie        Keir                   ckeir0@ucla.edu      F   
134  27515     Gaston  McCafferky         gmccafferky2d@reuters.com      M   
135  27479    Eustace     Chasson         echasson1d@technorati.com      M   
141  27500        Edi     Walczak      ewalczak1y@blogtalkradio.com      F   

     age   money         lon        lat       phone registration inactive  \
19   NaN  €83,79   30.156044 -23.411495         NaN   2017/06/06    False   
50   NaN  $23.07  -79.104850  -5.275574         NaN   2018/11/29    False   
59   NaN  $57.82    6.125060  45.925854         NaN   2018/08/13    False   
64   NaN  $95.86   31.887568  26.341876         NaN   2018/09/03    False   
71   NaN  $55.29  100.603798  16.158922         NaN   2019/04/22    False   
134  NaN     NaN  112.490900  -6.984200  0495501591   2018/11/04    False   
135  NaN     NaN   35.132780  31.423010         NaN   2019/07/27    False   
141  NaN  $73.80   20.231842  40.106379         NaN   2018/10/21    False   

        last_seen                           address           preference  \
19   1.524851e+09        Duiwelskloof, South Africa  entrée/plat/dessert   
50   1.543450e+09                    La Coipa, Peru  entrée/plat/dessert   
59   1.534118e+09                    Annecy, France  entrée/plat/dessert   
64   1.535933e+09                      Jirjā, Egypt  entrée/plat/dessert   
71   1.555891e+09               Thap Khlo, Thailand  entrée/plat/dessert   
134           NaN           Saniwonorejo, Indonesia  entrée/plat/dessert   
135           NaN  Al Karmil, Palestinian Territory  entrée/plat/dessert   
141  1.542229e+09             Saraqinishtë, Albania  entrée/plat/dessert   

             full_name          city                 country  
19       Tate Feldmann  Duiwelskloof            South Africa  
50        Hanny Ballin      La Coipa                    Peru  
59      Kinna Ffoulkes        Annecy                  France  
64     Brooke Edwicker         Jirjā                   Egypt  
71       Cordelie Keir     Thap Khlo                Thailand  
134  Gaston McCafferky  Saniwonorejo               Indonesia  
135    Eustace Chasson     Al Karmil   Palestinian Territory  
141        Edi Walczak  Saraqinishtë                 Albania

#### Traitement des dates

In [124]:
df['registration']

0      2019/04/16
1      2018/10/23
2      2018/12/30
3      2019/03/12
4      2018/08/13
          ...    
204    2019/06/04
205    2017/11/22
206    2018/08/20
207    2018/10/25
208    2017/06/05
Name: registration, Length: 203, dtype: object

In [125]:
df['registration'] = pd.to_datetime(df.registration)

<ipython-input-125-7108d870354f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['registration'] = pd.to_datetime(df.registration)


In [126]:
df['registration']

0     2019-04-16
1     2018-10-23
2     2018-12-30
3     2019-03-12
4     2018-08-13
         ...    
204   2019-06-04
205   2017-11-22
206   2018-08-20
207   2018-10-25
208   2017-06-05
Name: registration, Length: 203, dtype: datetime64[ns]

In [127]:
df.head()

id first_name  last_name                         email gender   age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1  27570      Ruthi       Ross              rross3w@sohu.com      F  57.0   
2  27572      Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4  27558    Margaux   Gowanson                nope@thankyou.      F  54.0   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019-04-16    False   
1  $20.37   19.230220  50.466575         NaN   2018-10-23    False   
2  €32,99  118.831081  24.984813         NaN   2018-12-30    False   
3  €98,93  121.648987  -8.844744         NaN   2019-03-12    False   
4  $13.30   14.772557  45.160472         NaN   2018-08-13    False   

      last_seen                    address           preference  \
0  1.559566e+09  Palkovice, Czech Republic  entrée/plat/dessert   
1  1.567165e+09           Siewierz, Poland  entrée/plat/dessert   
2  1.553692e+09              Longbo, China  entrée/plat/dessert   
3  1.552349e+09       Potulando, Indonesia          entrée/plat   
4  1.543383e+09            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country  
0   Leandra Pabelik  Palkovice   Czech Republic  
1        Ruthi Ross   Siewierz           Poland  
2    Silas Stourton     Longbo            China  
3  Roxine Pettecrew  Potulando        Indonesia  
4  Margaux Gowanson     Bribir          Croatia

In [128]:
df['last_seen'] = pd.to_datetime(df.last_seen, unit = 's')
df.head()

<ipython-input-128-21e1d01ef475>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['last_seen'] = pd.to_datetime(df.last_seen, unit = 's')


id first_name  last_name                         email gender   age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1  27570      Ruthi       Ross              rross3w@sohu.com      F  57.0   
2  27572      Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4  27558    Margaux   Gowanson                nope@thankyou.      F  54.0   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019-04-16    False   
1  $20.37   19.230220  50.466575         NaN   2018-10-23    False   
2  €32,99  118.831081  24.984813         NaN   2018-12-30    False   
3  €98,93  121.648987  -8.844744         NaN   2019-03-12    False   
4  $13.30   14.772557  45.160472         NaN   2018-08-13    False   

            last_seen                    address           preference  \
0 2019-06-03 12:39:57  Palkovice, Czech Republic  entrée/plat/dessert   
1 2019-08-30 11:41:04           Siewierz, Poland  entrée/plat/dessert   
2 2019-03-27 13:10:32              Longbo, China  entrée/plat/dessert   
3 2019-03-12 00:00:00       Potulando, Indonesia          entrée/plat   
4 2018-11-28 05:24:30            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country  
0   Leandra Pabelik  Palkovice   Czech Republic  
1        Ruthi Ross   Siewierz           Poland  
2    Silas Stourton     Longbo            China  
3  Roxine Pettecrew  Potulando        Indonesia  
4  Margaux Gowanson     Bribir          Croatia

In [130]:
pd.Series.fillna?

#### Traitement de la monnaie

In [133]:
np.random.seed(0) # pour avoir le même échantillon

In [134]:
df.sample(10)

id first_name   last_name                          email gender   age  \
18   27513   Dominick  Hryncewicz    dhryncewicz2b@amazonaws.com      M  56.0   
45   27432   Gardener     Kempson             gkempson2@furl.net      M  33.0   
33   27463     Caesar  Ferrarotti            cferrarottix@qq.com      M  59.0   
37   27617     Smitty     Richter       srichter57@pinterest.com      M  54.0   
112  27544  Cristiano    Midghall            cmidghall36@npr.org      M  47.0   
91   27618   Murielle      Beston            mbeston58@youku.com      F  38.0   
5    27497     Gordie     Bodicum          gbodicum1v@apache.org      M  60.0   
127  27431   Sinclair    Witherow          switherow1@joomla.org      M  41.0   
12   27548    Stanley    McCullen  smccullen3a@timesonline.co.uk      M  60.0   
157  27526   Sarajane      Fidler        sfidler2o@sitemeter.com      F  70.0   

      money         lon        lat       phone registration inactive  \
18   $80.71   68.214686  38.530112         NaN   2017-08-04    False   
45   $41.30   25.911060  49.103782  0774378059   2017-09-08    False   
33   €60.84  112.867154  -6.921845  0346857462   2018-12-10    False   
37   $19.07   37.149279  56.747788        0789   2018-10-20    False   
112  €43.94  -75.649902   4.827161  0574769298   2018-01-25    False   
91   €99,47  173.138978   1.386973         NaN   2017-02-19    False   
5    $47.26   13.978681  49.786243  0146943857   2017-04-29    False   
127  €55.95   22.004971  56.674534  0298040087   2018-12-26    False   
12      NaN  119.141473  28.074649  0875503094   2017-06-05    False   
157  $90.09   95.308054  32.596752  0511530696   2017-04-16    False   

              last_seen                    address           preference  \
18  2019-07-26 02:33:01     Tursunzoda, Tajikistan  entrée/plat/dessert   
45  2019-03-18 18:58:59       Kopychyntsi, Ukraine  entrée/plat/dessert   
33  2019-10-02 02:43:57    Trogan Barat, Indonesia  entrée/plat/dessert   
37  2019-04-14 11:57:19              Zarya, Russia  entrée/plat/dessert   
112 2019-06-15 14:02:49    Dos Quebradas, Colombia  entrée/plat/dessert   
91  2018-09-10 16:16:38  Bonriki Village, Kiribati  entrée/plat/dessert   
5   2018-12-26 07:40:36      Jince, Czech Republic  entrée/plat/dessert   
127 2018-12-26 00:00:00            Skrunda, Latvia  entrée/plat/dessert   
12                  NaT            Longquan, China  entrée/plat/dessert   
157 2018-08-09 11:29:14              Jieduo, China  entrée/plat/dessert   

               full_name             city          country  
18   Dominick Hryncewicz       Tursunzoda       Tajikistan  
45      Gardener Kempson      Kopychyntsi          Ukraine  
33     Caesar Ferrarotti     Trogan Barat        Indonesia  
37        Smitty Richter            Zarya           Russia  
112   Cristiano Midghall    Dos Quebradas         Colombia  
91       Murielle Beston  Bonriki Village         Kiribati  
5         Gordie Bodicum            Jince   Czech Republic  
127    Sinclair Witherow          Skrunda           Latvia  
12      Stanley McCullen         Longquan            China  
157      Sarajane Fidler           Jieduo            China

On a une colonne avec plusieurs devises. 

Produire une nouvelle colonne numérique 'money_eur'.
Pour la conversion USD/EUR, on utilise l'API https://api.exchangeratesapi.io/latest

In [135]:
# API
import json

import requests

response = requests.get('https://api.exchangeratesapi.io/latest')
rates = json.loads(response.content)
rates

{'rates': {'CAD': 1.5497,
  'HKD': 9.0993,
  'ISK': 163.2,
  'PHP': 57.148,
  'DKK': 7.442,
  'HUF': 364.36,
  'CZK': 27.279,
  'AUD': 1.6555,
  'RON': 4.8763,
  'SEK': 10.3593,
  'IDR': 17317.86,
  'INR': 86.0715,
  'BRL': 6.6054,
  'RUB': 91.5767,
  'HRK': 7.582,
  'JPY': 123.59,
  'THB': 36.608,
  'CHF': 1.0721,
  'SGD': 1.5942,
  'PLN': 4.5558,
  'BGN': 1.9558,
  'TRY': 9.33,
  'CNY': 7.8639,
  'NOK': 10.957,
  'NZD': 1.7746,
  'ZAR': 19.3867,
  'USD': 1.1741,
  'MXN': 24.872,
  'ILS': 3.9673,
  'GBP': 0.90915,
  'KRW': 1341.16,
  'MYR': 4.8708},
 'base': 'EUR',
 'date': '2020-10-16'}

In [136]:
# on va traiter la colonne money
df['money'].value_counts()

$42.44    3
$38.65    1
€38,83    1
€27,33    1
€33,95    1
         ..
$40.63    1
$98.26    1
€88,08    1
$53.88    1
$57.49    1
Name: money, Length: 184, dtype: int64

In [138]:
# il faut faire sauter le premier caractère qui correspond à la devise. 

In [139]:
df['currency'] = df['money'].str[0].map({'$' : 'USD', '€' : 'EUR'})

<ipython-input-139-2393c0158baf>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['currency'] = df['money'].str[0].map({'$' : 'USD', '€' : 'EUR'})


In [140]:
df.head()

id first_name  last_name                         email gender   age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1  27570      Ruthi       Ross              rross3w@sohu.com      F  57.0   
2  27572      Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4  27558    Margaux   Gowanson                nope@thankyou.      F  54.0   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019-04-16    False   
1  $20.37   19.230220  50.466575         NaN   2018-10-23    False   
2  €32,99  118.831081  24.984813         NaN   2018-12-30    False   
3  €98,93  121.648987  -8.844744         NaN   2019-03-12    False   
4  $13.30   14.772557  45.160472         NaN   2018-08-13    False   

            last_seen                    address           preference  \
0 2019-06-03 12:39:57  Palkovice, Czech Republic  entrée/plat/dessert   
1 2019-08-30 11:41:04           Siewierz, Poland  entrée/plat/dessert   
2 2019-03-27 13:10:32              Longbo, China  entrée/plat/dessert   
3 2019-03-12 00:00:00       Potulando, Indonesia          entrée/plat   
4 2018-11-28 05:24:30            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country currency  
0   Leandra Pabelik  Palkovice   Czech Republic      USD  
1        Ruthi Ross   Siewierz           Poland      USD  
2    Silas Stourton     Longbo            China      EUR  
3  Roxine Pettecrew  Potulando        Indonesia      EUR  
4  Margaux Gowanson     Bribir          Croatia      USD

In [141]:
df['money_eur'] = df['money'].str[1:].str.replace(',', '.')  # extraction des derniers chars + , => .
df['money_eur'] = pd.to_numeric(df['money_eur'])  # conversion en nombre

<ipython-input-141-01181753671c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['money_eur'] = df['money'].str[1:].str.replace(',', '.')  # extraction des derniers chars + , => .
<ipython-input-141-01181753671c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['money_eur'] = pd.to_numeric(df['money_eur'])  # conversion en nombre


In [142]:
df.head() # attention ici, la monnaie n'est pas encore convertie

id first_name  last_name                         email gender   age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1  27570      Ruthi       Ross              rross3w@sohu.com      F  57.0   
2  27572      Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4  27558    Margaux   Gowanson                nope@thankyou.      F  54.0   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019-04-16    False   
1  $20.37   19.230220  50.466575         NaN   2018-10-23    False   
2  €32,99  118.831081  24.984813         NaN   2018-12-30    False   
3  €98,93  121.648987  -8.844744         NaN   2019-03-12    False   
4  $13.30   14.772557  45.160472         NaN   2018-08-13    False   

            last_seen                    address           preference  \
0 2019-06-03 12:39:57  Palkovice, Czech Republic  entrée/plat/dessert   
1 2019-08-30 11:41:04           Siewierz, Poland  entrée/plat/dessert   
2 2019-03-27 13:10:32              Longbo, China  entrée/plat/dessert   
3 2019-03-12 00:00:00       Potulando, Indonesia          entrée/plat   
4 2018-11-28 05:24:30            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country currency  money_eur  
0   Leandra Pabelik  Palkovice   Czech Republic      USD      55.18  
1        Ruthi Ross   Siewierz           Poland      USD      20.37  
2    Silas Stourton     Longbo            China      EUR      32.99  
3  Roxine Pettecrew  Potulando        Indonesia      EUR      98.93  
4  Margaux Gowanson     Bribir          Croatia      USD      13.30

In [145]:
#la clé EUR = 1 # ie. conversion de l'eur par rapport à lui même

# conversion des monnaies en euros
rates['rates']['EUR'] = 1.0
df['money_eur'] = df['money_eur'] / df['currency'].map(rates['rates'])
np.random.seed(0)
df.sample(10)

<ipython-input-145-557d31e64583>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['money_eur'] = df['money_eur'] / df['currency'].map(rates['rates'])


id first_name   last_name                          email gender   age  \
18   27513   Dominick  Hryncewicz    dhryncewicz2b@amazonaws.com      M  56.0   
45   27432   Gardener     Kempson             gkempson2@furl.net      M  33.0   
33   27463     Caesar  Ferrarotti            cferrarottix@qq.com      M  59.0   
37   27617     Smitty     Richter       srichter57@pinterest.com      M  54.0   
112  27544  Cristiano    Midghall            cmidghall36@npr.org      M  47.0   
91   27618   Murielle      Beston            mbeston58@youku.com      F  38.0   
5    27497     Gordie     Bodicum          gbodicum1v@apache.org      M  60.0   
127  27431   Sinclair    Witherow          switherow1@joomla.org      M  41.0   
12   27548    Stanley    McCullen  smccullen3a@timesonline.co.uk      M  60.0   
157  27526   Sarajane      Fidler        sfidler2o@sitemeter.com      F  70.0   

      money         lon        lat       phone registration inactive  \
18   $80.71   68.214686  38.530112         NaN   2017-08-04    False   
45   $41.30   25.911060  49.103782  0774378059   2017-09-08    False   
33   €60.84  112.867154  -6.921845  0346857462   2018-12-10    False   
37   $19.07   37.149279  56.747788        0789   2018-10-20    False   
112  €43.94  -75.649902   4.827161  0574769298   2018-01-25    False   
91   €99,47  173.138978   1.386973         NaN   2017-02-19    False   
5    $47.26   13.978681  49.786243  0146943857   2017-04-29    False   
127  €55.95   22.004971  56.674534  0298040087   2018-12-26    False   
12      NaN  119.141473  28.074649  0875503094   2017-06-05    False   
157  $90.09   95.308054  32.596752  0511530696   2017-04-16    False   

              last_seen                    address           preference  \
18  2019-07-26 02:33:01     Tursunzoda, Tajikistan  entrée/plat/dessert   
45  2019-03-18 18:58:59       Kopychyntsi, Ukraine  entrée/plat/dessert   
33  2019-10-02 02:43:57    Trogan Barat, Indonesia  entrée/plat/dessert   
37  2019-04-14 11:57:19              Zarya, Russia  entrée/plat/dessert   
112 2019-06-15 14:02:49    Dos Quebradas, Colombia  entrée/plat/dessert   
91  2018-09-10 16:16:38  Bonriki Village, Kiribati  entrée/plat/dessert   
5   2018-12-26 07:40:36      Jince, Czech Republic  entrée/plat/dessert   
127 2018-12-26 00:00:00            Skrunda, Latvia  entrée/plat/dessert   
12                  NaT            Longquan, China  entrée/plat/dessert   
157 2018-08-09 11:29:14              Jieduo, China  entrée/plat/dessert   

               full_name             city          country currency  money_eur  
18   Dominick Hryncewicz       Tursunzoda       Tajikistan      USD  68.742015  
45      Gardener Kempson      Kopychyntsi          Ukraine      USD  35.175879  
33     Caesar Ferrarotti     Trogan Barat        Indonesia      EUR  60.840000  
37        Smitty Richter            Zarya           Russia      USD  16.242228  
112   Cristiano Midghall    Dos Quebradas         Colombia      EUR  43.940000  
91       Murielle Beston  Bonriki Village         Kiribati      EUR  99.470000  
5         Gordie Bodicum            Jince   Czech Republic      USD  40.252108  
127    Sinclair Witherow          Skrunda           Latvia      EUR  55.950000  
12      Stanley McCullen         Longquan            China      NaN        NaN  
157      Sarajane Fidler           Jieduo            China      USD  76.731113

#### Analyse des emails 

In [148]:
df['email'].isnull().sum() # combien d'email null ? 

4

In [149]:
len(df)

203

In [150]:
df = df.dropna(subset = ['email']).copy()

In [151]:
len(df)

199

In [152]:
#caractères non admis dans les emails 

# emails avec chars non admis
df.loc[df['email'].str.contains('[^A-Za-z0-9_\-%+.@]'), 'email'].values

array(['ycurston4a@nifty.com  ', '  agilmartin4b@un.org ',
       '  swinchcum4o@apple.com', '  erowcliffe43@chronoengine.com',
       'rbellenie54@topsy.com ', 'dbrindle1p@trellian.com ',
       ' efeatonby4i@hc360.com', ' mlowcock4u@nba.com'], dtype=object)

In [153]:
#nettoyage des emails pour enlever les blancs de part et d'autre de l'adresse mail 
df['email'] = df['email'].str.strip()
df.loc[df['email'].str.contains('[^A-Za-z0-9_\-%+.@]'), 'email'].values

array([], dtype=object)

In [154]:
with open('people.csv', encoding = 'latin-1') as f : 
    for line in f:
        print(line)
        break

id,first_name,last_name,email address,gender,age,money,lon,lat,phone,registration,inactive,last_seen,address,preference



In [155]:
line

'id,first_name,last_name,email address,gender,age,money,lon,lat,phone,registration,inactive,last_seen,address,preference\n'

In [156]:
line.rstrip() #vire le caractère \n

'id,first_name,last_name,email address,gender,age,money,lon,lat,phone,registration,inactive,last_seen,address,preference'

In [157]:
line

'id,first_name,last_name,email address,gender,age,money,lon,lat,phone,registration,inactive,last_seen,address,preference\n'

In [158]:
line.strip()

'id,first_name,last_name,email address,gender,age,money,lon,lat,phone,registration,inactive,last_seen,address,preference'

In [159]:
# regex pour vérifier les domaines
df.loc[~df['email'].str.contains('.+@[A-Za-z0-9_\-.]+\.[A-Za-z]{2,}')]

id first_name   last_name             email gender   age   money  \
4    27558    Margaux    Gowanson    nope@thankyou.      F  54.0  $13.30   
35   27560     Vinnie  Stansfield         non@merci      F  35.0  $21.85   
43   27575     Kaycee    Geillier  k.geillier@gmail      F  48.0  $53.88   
78   27561      Peggy    Scuffham        nein@danke      F  38.0  $43.86   
113  27562      Maure       Cronk         no@thanks      F  41.0  €57,41   

            lon        lat       phone registration inactive  \
4     14.772557  45.160472         NaN   2018-08-13    False   
35   -64.686609  10.190620         NaN   2018-04-16    False   
43    95.987223   5.246361         NaN   2018-05-18    False   
78   125.502581  11.501380  0511039244   2017-04-08    False   
113  111.856859  37.242565  0385826731   2017-04-26    False   

              last_seen                 address           preference  \
4   2018-11-28 05:24:30         Bribir, Croatia  entrée/plat/dessert   
35  2019-06-27 13:30:18    Lecherías, Venezuela  entrée/plat/dessert   
43  2018-07-10 23:58:03        Baroh, Indonesia  entrée/plat/dessert   
78  2018-08-20 16:41:00  Maydolong, Philippines  entrée/plat/dessert   
113 2018-12-21 12:46:48            Sanxi, China  entrée/plat/dessert   

             full_name       city       country currency  money_eur  
4     Margaux Gowanson     Bribir       Croatia      USD  11.327826  
35   Vinnie Stansfield  Lecherías     Venezuela      USD  18.609999  
43     Kaycee Geillier      Baroh     Indonesia      USD  45.890469  
78      Peggy Scuffham  Maydolong   Philippines      USD  37.356273  
113        Maure Cronk      Sanxi         China      EUR  57.410000

In [166]:
df.loc[df['email'].str.contains('.+@[A-Za-z0-9_\-.]+\.[A-Za-z]{2,}')]

id  first_name  last_name                         email gender   age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1    27570       Ruthi       Ross              rross3w@sohu.com      F  57.0   
2    27572       Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
5    27497      Gordie    Bodicum         gbodicum1v@apache.org      M  60.0   
..     ...         ...        ...                           ...    ...   ...   
204  27465     Adelind   Christin          achristinz@blogs.com      F  30.0   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu      F  28.0   
206  27521      Moritz    Issacof          missacof2j@wired.com      M  21.0   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com      F  32.0   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl      M  34.0   

      money         lon        lat       phone registration inactive  \
0    $55.18   18.284100  49.632552  0136319724   2019-04-16    False   
1    $20.37   19.230220  50.466575         NaN   2018-10-23    False   
2    €32,99  118.831081  24.984813         NaN   2018-12-30    False   
3    €98,93  121.648987  -8.844744         NaN   2019-03-12    False   
5    $47.26   13.978681  49.786243  0146943857   2017-04-29    False   
..      ...         ...        ...         ...          ...      ...   
204  €81,84   41.427853  52.415968         NaN   2019-06-04    False   
205  €37.15  132.702111  33.762296         NaN   2017-11-22    False   
206  $91.97   72.311099  31.263396  0298949280   2018-08-20    False   
207  €55,64  111.138442  -6.783653  0734595126   2018-10-25    False   
208  $81.54  -47.208953 -23.088650  0516973570   2017-06-05    False   

              last_seen                     address           preference  \
0   2019-06-03 12:39:57   Palkovice, Czech Republic  entrée/plat/dessert   
1   2019-08-30 11:41:04            Siewierz, Poland  entrée/plat/dessert   
2   2019-03-27 13:10:32               Longbo, China  entrée/plat/dessert   
3   2019-03-12 00:00:00        Potulando, Indonesia          entrée/plat   
5   2018-12-26 07:40:36       Jince, Czech Republic  entrée/plat/dessert   
..                  ...                         ...                  ...   
204 2019-06-04 00:00:00            Znamenka, Russia  entrée/plat/dessert   
205 2018-07-29 13:18:29                  Iyo, Japan  entrée/plat/dessert   
206 2019-02-10 16:39:03        Jhang Sadr, Pakistan  entrée/plat/dessert   
207 2018-10-25 00:00:00  Mantingantengah, Indonesia  entrée/plat/dessert   
208 2018-10-25 04:20:32          Indaiatuba, Brazil  entrée/plat/dessert   

               full_name             city          country currency  money_eur  
0        Leandra Pabelik        Palkovice   Czech Republic      USD  46.997700  
1             Ruthi Ross         Siewierz           Poland      USD  17.349459  
2         Silas Stourton           Longbo            China      EUR  32.990000  
3       Roxine Pettecrew        Potulando        Indonesia      EUR  98.930000  
5         Gordie Bodicum            Jince   Czech Republic      USD  40.252108  
..                   ...              ...              ...      ...        ...  
204     Adelind Christin         Znamenka           Russia      EUR  81.840000  
205  Petronella Pickance              Iyo            Japan      EUR  37.150000  
206       Moritz Issacof       Jhang Sadr         Pakistan      USD  78.332340  
207   Quintana Foulstone  Mantingantengah        Indonesia      EUR  55.640000  
208     Trueman De Haven       Indaiatuba           Brazil      USD  69.448940  

[194 rows x 20 columns]

Attention, il faut mettre le tilde pour avoir les emails qui ne sont pas au même format que celui du regex.

le regex ici représente : '.+@[A-Za-z0-9_\-.]+\.[A-Za-z]{2,}'

ici, le regex nous donne les emails valides avec @"caractères"."2 lettres minimum"

le tilde nous donne le contraire de l'ensemble identifié par le regex

Pour les séries de pandas, on utilise : 
    - tilde pour la négation 
    - & pour et
    - | pour ou 

In [165]:
#email avec des aliases 

df.loc[df['email'].str.contains('\+'), 'email'].values

array([], dtype=object)

In [162]:
#pour supprimer les alias 

tab_email = df['email'].str.extract('([^+]+)(\+.*)?(@.+)').sort_values(0)
tab_email

0       1                     2
89   Izzy.Clemence45     NaN              @hhs.gov
185           a.gorz     NaN            @gmail.com
186           a.gorz  +alias            @gmail.com
40     aMatthisson3s     NaN              @ted.com
39          abonar23     NaN            @wufoo.com
..               ...     ...                   ...
159     wconcannon2l     NaN  @merriam-webster.com
53        wshewery11     NaN          @smugmug.com
74        ycurston4a     NaN            @nifty.com
80       zreinhard2f     NaN           @github.com
195      zsmewings3j     NaN             @ucsd.edu

[199 rows x 3 columns]

ici on a : '([^+]+)(\+.*)?(@.+)'
    
avec: 
- '([^+]+)représentant extraction 0
- (\+.*)représentant extraction 1
- ?(@.+)' représentant extraction 2

In [163]:
df['email']=tab_email[0] + tab_email[2]

In [164]:
df.head()

id first_name  last_name                         email gender   age  \
0  27625    Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1  27570      Ruthi       Ross              rross3w@sohu.com      F  57.0   
2  27572      Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3  27435     Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4  27558    Margaux   Gowanson                nope@thankyou.      F  54.0   

    money         lon        lat       phone registration inactive  \
0  $55.18   18.284100  49.632552  0136319724   2019-04-16    False   
1  $20.37   19.230220  50.466575         NaN   2018-10-23    False   
2  €32,99  118.831081  24.984813         NaN   2018-12-30    False   
3  €98,93  121.648987  -8.844744         NaN   2019-03-12    False   
4  $13.30   14.772557  45.160472         NaN   2018-08-13    False   

            last_seen                    address           preference  \
0 2019-06-03 12:39:57  Palkovice, Czech Republic  entrée/plat/dessert   
1 2019-08-30 11:41:04           Siewierz, Poland  entrée/plat/dessert   
2 2019-03-27 13:10:32              Longbo, China  entrée/plat/dessert   
3 2019-03-12 00:00:00       Potulando, Indonesia          entrée/plat   
4 2018-11-28 05:24:30            Bribir, Croatia  entrée/plat/dessert   

          full_name       city          country currency  money_eur  
0   Leandra Pabelik  Palkovice   Czech Republic      USD  46.997700  
1        Ruthi Ross   Siewierz           Poland      USD  17.349459  
2    Silas Stourton     Longbo            China      EUR  32.990000  
3  Roxine Pettecrew  Potulando        Indonesia      EUR  98.930000  
4  Margaux Gowanson     Bribir          Croatia      USD  11.327826

In [167]:
df['email'] = tab_email[0] + tab_email[2]
df.sort_values('email')

id first_name   last_name                             email gender  \
89   27579       Izzy    Clemence           Izzy.Clemence45@hhs.gov      M   
185  27596      André        Gorz                  a.gorz@gmail.com      M   
186  27902      Goran        Drez                  a.gorz@gmail.com      M   
40   27566     Amalie  Matthisson             aMatthisson3s@ted.com      F   
39   27505      Alvis       Bonar                abonar23@wufoo.com      M   
..     ...        ...         ...                               ...    ...   
159  27523   Wolfgang   Concannon  wconcannon2l@merriam-webster.com      M   
53   27467     Worden     Shewery            wshewery11@smugmug.com      M   
74   27584    Yehudit     Curston              ycurston4a@nifty.com      M   
80   27517       Zack    Reinhard            zreinhard2f@github.com      M   
195  27557    Zachery    Smewings              zsmewings3j@ucsd.edu      M   

      age   money         lon        lat       phone registration inactive  \
89   48.0  €46.75   97.521763  40.324768         NaN   2019-08-10    False   
185  43.0  €17.20   88.166695  39.024258         NaN   2019-05-23    False   
186  55.0     NaN   88.166695  39.024258         NaN   2019-07-23    False   
40   63.0  €13,89   16.304150  57.993607  0409272683   2019-01-20    False   
39   27.0  $90.88   14.426644  49.660577         NaN   2017-03-09    False   
..    ...     ...         ...        ...         ...          ...      ...   
159  59.0  €51.68  120.174895  30.387552         NaN   2017-07-10    False   
53   26.0  $15.67   19.883333  40.916667  0517301167   2018-08-22    False   
74   63.0  $86.99 -106.417031  31.975476  0157438239   2019-01-09    False   
80   35.0  $91.20  -68.103824 -38.961513  0056565079   2017-10-20    False   
195  31.0  €65.32  112.999502  28.190502         NaN   2019-09-08    False   

              last_seen                   address           preference  \
89  2019-08-10 00:00:00             Xigebi, China  entrée/plat/dessert   
185 2019-10-11 20:05:21            Vosnon, France  entrée/plat/dessert   
186                 NaT             Paris, France  entrée/plat/dessert   
40  2019-01-20 00:00:00         Linköping, Sweden  entrée/plat/dessert   
39  2019-03-09 13:06:07  Sedlčany, Czech Republic  entrée/plat/dessert   
..                  ...                       ...                  ...   
159 2019-10-06 16:02:10          Chongxian, China  entrée/plat/dessert   
53  2018-11-22 07:43:27            Kajan, Albania  entrée/plat/dessert   
74  2019-01-09 00:00:00    El Paso, United States  entrée/plat/dessert   
80  2019-04-21 06:56:42      Las Lajas, Argentina  entrée/plat/dessert   
195 2019-09-08 00:00:00             Luowan, China  entrée/plat/dessert   

              full_name       city          country currency  money_eur  
89        Izzy Clemence     Xigebi            China      EUR  46.750000  
185          André Gorz     Vosnon           France      EUR  17.200000  
186          Goran Drez      Paris           France      NaN        NaN  
40    Amalie Matthisson  Linköping           Sweden      EUR  13.890000  
39          Alvis Bonar   Sedlčany   Czech Republic      USD  77.403969  
..                  ...        ...              ...      ...        ...  
159  Wolfgang Concannon  Chongxian            China      EUR  51.680000  
53       Worden Shewery      Kajan          Albania      USD  13.346393  
74      Yehudit Curston    El Paso    United States      USD  74.090793  
80        Zack Reinhard  Las Lajas        Argentina      USD  77.676518  
195    Zachery Smewings     Luowan            China      EUR  65.320000  

[199 rows x 20 columns]

In [168]:
df = df.drop_duplicates(subset=['email'])
df.sort_values('email')

id first_name   last_name                             email gender  \
89   27579       Izzy    Clemence           Izzy.Clemence45@hhs.gov      M   
185  27596      André        Gorz                  a.gorz@gmail.com      M   
40   27566     Amalie  Matthisson             aMatthisson3s@ted.com      F   
39   27505      Alvis       Bonar                abonar23@wufoo.com      M   
204  27465    Adelind    Christin              achristinz@blogs.com      F   
..     ...        ...         ...                               ...    ...   
159  27523   Wolfgang   Concannon  wconcannon2l@merriam-webster.com      M   
53   27467     Worden     Shewery            wshewery11@smugmug.com      M   
74   27584    Yehudit     Curston              ycurston4a@nifty.com      M   
80   27517       Zack    Reinhard            zreinhard2f@github.com      M   
195  27557    Zachery    Smewings              zsmewings3j@ucsd.edu      M   

      age   money         lon        lat       phone registration inactive  \
89   48.0  €46.75   97.521763  40.324768         NaN   2019-08-10    False   
185  43.0  €17.20   88.166695  39.024258         NaN   2019-05-23    False   
40   63.0  €13,89   16.304150  57.993607  0409272683   2019-01-20    False   
39   27.0  $90.88   14.426644  49.660577         NaN   2017-03-09    False   
204  30.0  €81,84   41.427853  52.415968         NaN   2019-06-04    False   
..    ...     ...         ...        ...         ...          ...      ...   
159  59.0  €51.68  120.174895  30.387552         NaN   2017-07-10    False   
53   26.0  $15.67   19.883333  40.916667  0517301167   2018-08-22    False   
74   63.0  $86.99 -106.417031  31.975476  0157438239   2019-01-09    False   
80   35.0  $91.20  -68.103824 -38.961513  0056565079   2017-10-20    False   
195  31.0  €65.32  112.999502  28.190502         NaN   2019-09-08    False   

              last_seen                   address           preference  \
89  2019-08-10 00:00:00             Xigebi, China  entrée/plat/dessert   
185 2019-10-11 20:05:21            Vosnon, France  entrée/plat/dessert   
40  2019-01-20 00:00:00         Linköping, Sweden  entrée/plat/dessert   
39  2019-03-09 13:06:07  Sedlčany, Czech Republic  entrée/plat/dessert   
204 2019-06-04 00:00:00          Znamenka, Russia  entrée/plat/dessert   
..                  ...                       ...                  ...   
159 2019-10-06 16:02:10          Chongxian, China  entrée/plat/dessert   
53  2018-11-22 07:43:27            Kajan, Albania  entrée/plat/dessert   
74  2019-01-09 00:00:00    El Paso, United States  entrée/plat/dessert   
80  2019-04-21 06:56:42      Las Lajas, Argentina  entrée/plat/dessert   
195 2019-09-08 00:00:00             Luowan, China  entrée/plat/dessert   

              full_name       city          country currency  money_eur  
89        Izzy Clemence     Xigebi            China      EUR  46.750000  
185          André Gorz     Vosnon           France      EUR  17.200000  
40    Amalie Matthisson  Linköping           Sweden      EUR  13.890000  
39          Alvis Bonar   Sedlčany   Czech Republic      USD  77.403969  
204    Adelind Christin   Znamenka           Russia      EUR  81.840000  
..                  ...        ...              ...      ...        ...  
159  Wolfgang Concannon  Chongxian            China      EUR  51.680000  
53       Worden Shewery      Kajan          Albania      USD  13.346393  
74      Yehudit Curston    El Paso    United States      USD  74.090793  
80        Zack Reinhard  Las Lajas        Argentina      USD  77.676518  
195    Zachery Smewings     Luowan            China      EUR  65.320000  

[196 rows x 20 columns]

#### Analyse de la colonne préférence 

In [169]:
df['preference'].value_counts()

entrée/plat/dessert    185
plat/dessert             4
plat                     3
entrée/plat              3
boisson                  1
Name: preference, dtype: int64

In [171]:
s = set()
df['preference'].apply(lambda x : s.update(x.split('/')))
s

{'boisson', 'dessert', 'entrée', 'plat'}

In [172]:
for x in sorted(s): 
    df[x] = df['preference'].str.contains(x)

df

id  first_name  last_name                         email gender   age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1    27570       Ruthi       Ross              rross3w@sohu.com      F  57.0   
2    27572       Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4    27558     Margaux   Gowanson                nope@thankyou.      F  54.0   
..     ...         ...        ...                           ...    ...   ...   
204  27465     Adelind   Christin          achristinz@blogs.com      F  30.0   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu      F  28.0   
206  27521      Moritz    Issacof          missacof2j@wired.com      M  21.0   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com      F  32.0   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl      M  34.0   

      money         lon        lat       phone  ...           preference  \
0    $55.18   18.284100  49.632552  0136319724  ...  entrée/plat/dessert   
1    $20.37   19.230220  50.466575         NaN  ...  entrée/plat/dessert   
2    €32,99  118.831081  24.984813         NaN  ...  entrée/plat/dessert   
3    €98,93  121.648987  -8.844744         NaN  ...          entrée/plat   
4    $13.30   14.772557  45.160472         NaN  ...  entrée/plat/dessert   
..      ...         ...        ...         ...  ...                  ...   
204  €81,84   41.427853  52.415968         NaN  ...  entrée/plat/dessert   
205  €37.15  132.702111  33.762296         NaN  ...  entrée/plat/dessert   
206  $91.97   72.311099  31.263396  0298949280  ...  entrée/plat/dessert   
207  €55,64  111.138442  -6.783653  0734595126  ...  entrée/plat/dessert   
208  $81.54  -47.208953 -23.088650  0516973570  ...  entrée/plat/dessert   

               full_name             city          country currency  \
0        Leandra Pabelik        Palkovice   Czech Republic      USD   
1             Ruthi Ross         Siewierz           Poland      USD   
2         Silas Stourton           Longbo            China      EUR   
3       Roxine Pettecrew        Potulando        Indonesia      EUR   
4       Margaux Gowanson           Bribir          Croatia      USD   
..                   ...              ...              ...      ...   
204     Adelind Christin         Znamenka           Russia      EUR   
205  Petronella Pickance              Iyo            Japan      EUR   
206       Moritz Issacof       Jhang Sadr         Pakistan      USD   
207   Quintana Foulstone  Mantingantengah        Indonesia      EUR   
208     Trueman De Haven       Indaiatuba           Brazil      USD   

     money_eur boisson dessert entrée  plat  
0    46.997700   False    True   True  True  
1    17.349459   False    True   True  True  
2    32.990000   False    True   True  True  
3    98.930000   False   False   True  True  
4    11.327826   False    True   True  True  
..         ...     ...     ...    ...   ...  
204  81.840000   False    True   True  True  
205  37.150000   False    True   True  True  
206  78.332340   False    True   True  True  
207  55.640000   False    True   True  True  
208  69.448940   False    True   True  True  

[196 rows x 24 columns]

In [175]:
#si utilisation d'une liste...  on retrouve les listes dans les banques 
# produit du client dans les listes. 
l = list()
df['preference'].apply(lambda x : l.extend(x.split('/')))
l

['entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 'plat',
 'dessert',
 'entrée',
 

In [176]:
df['preference'].str.get_dummies(sep = '/') # permet de faire le même boulot que précédemment

boisson  dessert  entrée  plat
0          0        1       1     1
1          0        1       1     1
2          0        1       1     1
3          0        0       1     1
4          0        1       1     1
..       ...      ...     ...   ...
204        0        1       1     1
205        0        1       1     1
206        0        1       1     1
207        0        1       1     1
208        0        1       1     1

[196 rows x 4 columns]

In [177]:
tab_preference = df['preference'].str.get_dummies(sep='/')
df[tab_preference.columns] = tab_preference.astype(bool)
df

id  first_name  last_name                         email gender   age  \
0    27625     Leandra    Pabelik           lpabelik5f@yale.edu      F  63.0   
1    27570       Ruthi       Ross              rross3w@sohu.com      F  57.0   
2    27572       Silas   Stourton  silas.stourton3y@answers.com      M  22.0   
3    27435      Roxine  Pettecrew      rpettecrew5@gravatar.com      F  64.0   
4    27558     Margaux   Gowanson                nope@thankyou.      F  54.0   
..     ...         ...        ...                           ...    ...   ...   
204  27465     Adelind   Christin          achristinz@blogs.com      F  30.0   
205  27476  Petronella   Pickance          ppickance1a@uiuc.edu      F  28.0   
206  27521      Moritz    Issacof          missacof2j@wired.com      M  21.0   
207  27441    Quintana  Foulstone      qfoulstoneb@newsvine.com      F  32.0   
208  27597     Trueman   De Haven             tdehaven4n@goo.gl      M  34.0   

      money         lon        lat       phone  ...           preference  \
0    $55.18   18.284100  49.632552  0136319724  ...  entrée/plat/dessert   
1    $20.37   19.230220  50.466575         NaN  ...  entrée/plat/dessert   
2    €32,99  118.831081  24.984813         NaN  ...  entrée/plat/dessert   
3    €98,93  121.648987  -8.844744         NaN  ...          entrée/plat   
4    $13.30   14.772557  45.160472         NaN  ...  entrée/plat/dessert   
..      ...         ...        ...         ...  ...                  ...   
204  €81,84   41.427853  52.415968         NaN  ...  entrée/plat/dessert   
205  €37.15  132.702111  33.762296         NaN  ...  entrée/plat/dessert   
206  $91.97   72.311099  31.263396  0298949280  ...  entrée/plat/dessert   
207  €55,64  111.138442  -6.783653  0734595126  ...  entrée/plat/dessert   
208  $81.54  -47.208953 -23.088650  0516973570  ...  entrée/plat/dessert   

               full_name             city          country currency  \
0        Leandra Pabelik        Palkovice   Czech Republic      USD   
1             Ruthi Ross         Siewierz           Poland      USD   
2         Silas Stourton           Longbo            China      EUR   
3       Roxine Pettecrew        Potulando        Indonesia      EUR   
4       Margaux Gowanson           Bribir          Croatia      USD   
..                   ...              ...              ...      ...   
204     Adelind Christin         Znamenka           Russia      EUR   
205  Petronella Pickance              Iyo            Japan      EUR   
206       Moritz Issacof       Jhang Sadr         Pakistan      USD   
207   Quintana Foulstone  Mantingantengah        Indonesia      EUR   
208     Trueman De Haven       Indaiatuba           Brazil      USD   

     money_eur boisson dessert entrée  plat  
0    46.997700   False    True   True  True  
1    17.349459   False    True   True  True  
2    32.990000   False    True   True  True  
3    98.930000   False   False   True  True  
4    11.327826   False    True   True  True  
..         ...     ...     ...    ...   ...  
204  81.840000   False    True   True  True  
205  37.150000   False    True   True  True  
206  78.332340   False    True   True  True  
207  55.640000   False    True   True  True  
208  69.448940   False    True   True  True  

[196 rows x 24 columns]